elki/src/main/java/de/lmu/ifi/dbs/elki/index/tree/spatial/ph/MinimalisticMemoryPHTree.java

```diff
@@ -0,0 +1,397 @@
+import de.lmu.ifi.dbs.elki.distance.distancefunction.Norm;
+import de.lmu.ifi.dbs.elki.distance.distancefunction.minkowski.LPNormDistanceFunction;
+import de.lmu.ifi.dbs.elki.distance.distancefunction.minkowski.SparseLPNormDistanceFunction;
+import de.lmu.ifi.dbs.elki.distance.distancefunction.minkowski.SquaredEuclideanDistanceFunction;
+import de.lmu.ifi.dbs.elki.index.AbstractIndex;
+import de.lmu.ifi.dbs.elki.index.DynamicIndex;
+import de.lmu.ifi.dbs.elki.index.IndexFactory;
+import de.lmu.ifi.dbs.elki.index.KNNIndex;
+import de.lmu.ifi.dbs.elki.index.RangeIndex;
+import de.lmu.ifi.dbs.elki.logging.Logging;
+import de.lmu.ifi.dbs.elki.logging.statistics.Counter;
+import de.lmu.ifi.dbs.elki.utilities.Alias;
+import de.lmu.ifi.dbs.elki.utilities.documentation.Reference;
+import de.lmu.ifi.dbs.elki.utilities.optionhandling.AbstractParameterizer;
+
+/**
+ * Implementation of an in-memory PH-tree. 
+ * 
+ * @author Tilmann Zaeschke, Erich Schubert
+ * 
+ * @apiviz.has PHTreeKNNQuery
+ * @apiviz.has PHTreeRangeQuery
+ * 
+ * @param <O> Vector type
+ */
+@Reference(authors = "T. Zaeschke, C. Zimmerli, M.C. Norrie", title = "The PH-Tree -- A Space-Efficient Storage Structure and Multi-Dimensional Index", booktitle = "Proc. Intl. Conf. on Management of Data (SIGMOD'14), 2014", url = "http://dx.doi.org/10.1145/361002.361007")
+public class MinimalisticMemoryPHTree<O extends NumberVector> extends AbstractIndex<O> 
+    implements DynamicIndex,  KNNIndex<O>, RangeIndex<O> {
+  /**
+   * Class logger
+   */
+  private static final Logging LOG = Logging.getLogger(MinimalisticMemoryPHTree.class);
+
+  
+  
+  /**
+   * The actual "tree" as a sorted array.
+   */
+  ArrayModifiableDBIDs sorted = null;
+  
+  private final PhTreeF<DBID> tree;
+
+  /**
+   * The number of dimensions.
+   */
+  private int dims = -1;
+
+  /**
+   * Counter for comparisons.
+   */
+  private final Counter objaccess;
+
+  /**
+   * Counter for distance computations.
+   */
+  private final Counter distcalc;
+
+  /**
+   * Constructor.
+   * 
+   * @param relation Relation to index
+   */
+  public MinimalisticMemoryPHTree(Relation<O> relation) {
+    super(relation);
+    if(LOG.isStatistics()) {
+      String prefix = this.getClass().getName();
+      this.objaccess = LOG.newCounter(prefix + ".objaccess");
+      this.distcalc = LOG.newCounter(prefix + ".distancecalcs");
+    }
+    else {
+      this.objaccess = null;
+      this.distcalc = null;
+    }
+    dims = RelationUtil.dimensionality(relation);
+    tree = PhTreeF.create(dims);
+  }
+
+  @Override
+  public void initialize() {
+    sorted = DBIDUtil.newArray(relation.getDBIDs());
+
+
+    DBIDIter iter = relation.getDBIDs().iter();
+
+    for(; iter.valid(); iter.advance()) {
+      O o = relation.get(iter);
+      double[] v = new double[dims];
+      for (int k = 0; k < dims; k++) {
+        v[k] = o.doubleValue(k);
+      }
+      DBID id = DBIDUtil.deref(iter);
```

Comment:
If this is about the `sorted` field, I removed it, it was not necessary.

If it is about the loop: Probably I haven't really understood the design of ELKI yet. Could you show me how to access the data without materializing all the DBIDs?

Also, for the query results I need to return the database identifiers, I think. Since the PhTree only accepts Objects as values, I could Store an `Integer` object instead of the `DBID` objects, but I think that would not help much.

Do you have any suggestions? Did I understand you r comment correctly?